### PR TITLE
Add events page

### PR DESCRIPTION
**Pull request by alexdmtr**
- All events scraped from Facebook Graph API
- All events linked to their Facebook pages
- All events cached, requests made under 1 minute intervals are instant
- Add error checking to rendering templates
- Add loading environment variables from a single .env file

### CODE REVIEW

File: app.go [hunk 1]
```diff
@@ -6,7 +6,8 @@ import (
 	"os"
 	"path/filepath"
 
-	"github.com/hacksoc-manchester/www/handlers"
+	_ "github.com/alexdmtr/www/config"
+	"github.com/alexdmtr/www/handlers"
```
Comment (rignatus) [inline app.go]: These paths should be updated back to "github.com/hacksoc-manchester/www/".

File: app.go [hunk 2]
```diff
@@ -22,10 +23,15 @@ func main() {
...
-		http.ListenAndServe(":"+os.Getenv("HTTP_PLATFORM_PORT"), nil)
+		err = http.ListenAndServe(":"+os.Getenv("HTTP_PLATFORM_PORT"), nil)
 	} else {
-		http.ListenAndServe(":8080", nil)
+		err = http.ListenAndServe(":8080", nil)
+	}
+
+	if err != nil {
+		log.Fatal(err)
 	}
```
Comment (rignatus) [inline app.go]: Maybe we could change every `http.ListenAndServe()` call to `log.Fatal(ListenAndServe())`?

File: assets/css/events.css [hunk 3]
```diff
@@ -1,118 +1,113 @@
-/* Widescreen */
-#events ul {
-    padding: 30px 130px; padding: 30px 18vw;
+.event-group {
+    /*padding: 30px 140px;*/
```
Comment (MisterDoom) [inline assets/css/events.css]: Nit: we could remove this comment.

File: config/config.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+import (
+	"github.com/joho/godotenv"
+	"log"
+)
+
+func init() {
+	err := godotenv.Load()
+	if err != nil {
+		log.Fatal(err)
+		log.Fatal("Error loading .env file")
```
Comment (rignatus) [inline config/config.go]: This error message is redundant as `err` already contains enough information. 

Moreover, this line will never be reached as the previous `log.Fatal(err)` calls `os.Exit(1)`. A possible solution could be `log.Fatal("Error load .env file ", err)`.

File: handlers/error.go [hunk 5]
```diff
@@ -1,13 +1,25 @@
...
+		message = messageModel{"Error", "Page not found."}
+	case http.StatusServiceUnavailable:
+		message = messageModel{"Error", "Page not available at this time. Try again later."}
+	default:
+		message = messageModel{"Error", "An error has occurred."}
+	}
+
+	err := templates["message"].ExecuteTemplate(w, "layout", message)
+	if err != nil {
+		log.Println(err)
```
Comment (rignatus) [inline handlers/error.go:23]: How about changing this:
``` go
err := templates["message"].ExecuteTemplate(w, "layout", message)
if err != nil {
	log.Println(err)
```
to
``` go

if err := templates["message"].ExecuteTemplate(w, "layout", message); err != nil {
	log.Println(err)
```
?

File: handlers/events.go [hunk 6]
```diff
@@ -1,10 +1,33 @@
...
+		HaveRightNow bool
+		HaveUpcoming bool
+	}
+	eventGroup, err := eventService.GroupEvents()
+
+	if err != nil {
+		log.Println(err)
+		errorHandler(w, r, http.StatusServiceUnavailable)
+		return
+	}
```
Comment (rignatus) [inline handlers/events.go:23]: Very nice way of handling the error as well as displaying something to the user!

File: handlers/events.go [hunk 7]
```diff
@@ -1,10 +1,33 @@
...
+
+	eventsContext.EventGroup = eventGroup
+	eventsContext.HaveRightNow = len(eventGroup.RightNow) > 0
+	eventsContext.HaveUpcoming = len(eventGroup.Upcoming) > 0
+
+	err = templates["events"].ExecuteTemplate(w, "layout", eventsContext)
+
+	if err != nil {
+		log.Println(err)
+	}
```
Comment (rignatus) [inline handlers/events.go]: Could make the same change as in `error.go`, suggested above.

File: services/eventService/eventService.go [hunk 8]
```diff
@@ -1,43 +1,250 @@
...
+	Location        string
+	URL             string
+	ImageURL        string
+	AttendingCount  int
+	InterestedCount int
+	StartTime       time.Time
+	EndTime         time.Time
+}
+
+func (e *Event) GetInterestLine() string {
```
Comment (rignatus) [inline services/eventService/eventService.go:29]: How about `GetParticipation` or `GetInterestCount`?

File: services/eventService/eventService.go [hunk 9]
```diff
@@ -1,43 +1,250 @@
...
+			"fields": "name,description,cover,place,start_time,end_time,attending_count,interested_count",
+		}, true)
+
+	client := &http.Client{}
+	response, err := client.Do(req)
+
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (rignatus) [inline services/eventService/eventService.go:110]: This error handling is not complete as `client.Do` will not cause an error if there is a non-2xx status. Another if, checking for a non-2xx status is required.

File: handlers/error.go (hunk 5 above)
Comment (alexdmtr) [reply handlers/error.go:23]: Line too long in my opinon.
But I have a solution:
- create renderTemplate() func
- call that instead of manually executing templates
- respects DRY
- can handle errors appropriately there

File: app.go (hunk 2 above)
Comment (alexdmtr) [reply app.go]: Agree

File: handlers/events.go (hunk 6 above)
Comment (alexdmtr) [reply handlers/events.go:23]: Thanks!

File: services/eventService/eventService.go [hunk 10]
```diff
@@ -1,43 +1,250 @@
...
+			if e.StartTime.Month() == e.EndTime.Month() {
+				return fmt.Sprintf("%s - %s", startTime, e.EndTime.Format("2"))
+			}
+			var endTime string
+
+			if e.EndTime.Year() == time.Now().Year() {
+				endTime = e.EndTime.Format("January 2")
+			} else {
+				endTime = e.EndTime.Format("January 2, 2006")
+			}
```
Comment (rignatus) [inline services/eventService/eventService.go]: How about changing 
``` go
var endTime string
if e.EndTime.Year() == time.Now().Year() {
	endTime = e.EndTime.Format("January 2")
} else {
	endTime = e.EndTime.Format("January 2, 2006")
}
```
to
``` go
endTime := e.EndTime.Format("January 2, 2006")
if e.EndTime.Year() == time.Now().Year() {
	endTime = e.EndTime.Format("January 2")
}
```
?

File: services/eventService/eventService.go [hunk 11]
```diff
@@ -1,43 +1,250 @@
...
+	if os.Getenv("FB_APP_ID") == "" {
+		log.Println("Environment variable FB_APP_ID is not assigned.")
+		return
 	}
 
-	return events
+	if os.Getenv("FB_SECRET") == "" {
+		log.Println("Environment variable FB_SECRET is not assigned.")
+		return
+	}
```
Comment (rignatus) [inline services/eventService/eventService.go]: Nit: if I won't have both "FB_APP_ID" and "FB_SECRET", I won't be notified about the latter.

File: templates/shared/event-list.tmpl [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+                <a href="{{ .URL }}" class="event-box">
+                    <div class="event-figure">
+                            <img src="{{.ImageURL}}" class="event-img"/>
+                    </div>
+                    <div class="event-details">
+                            <h2 class="event-date">{{ .GetShortDate }}</h2>
+
+                            <h3 class="event-name">{{ .Name }}</h3>
+
+
```
Comment (rignatus) [inline templates/shared/event-list.tmpl]: Nit: remove empty lines.

File: templates/shared/event-list.tmpl [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+                            <img src="assets/images/social-media/facebook-grey.png" alt="Facebook" class="event-facebook">
+                          </span>
+                    </div>
+                </a>
+            </div>
+
+        {{end}}
+    </div>
+
+{{end}}
```
Comment (rignatus) [inline templates/shared/event-list.tmpl]: As we are using 2 spaces tabs along our templates, It is a good idea to maintain consistency in tabs-length.

File: assets/css/events.css (hunk 3 above)
Comment (rignatus) [inline assets/css/events.css]: Is this comment having any purpose?

File: assets/css/events.css [hunk 14]
```diff
@@ -1,118 +1,113 @@
...
+    justify-content: center;
+    padding: 0 15px 0 15px;
+}
+
+.event {
+    width: 300px;
+    height: 300px;
+    margin: 15px;
+    /*box-shadow: 0 1px 2px 0 rgba(0,0,0,.15);*/
+    box-shadow: 0 1px 0 0 rgba(0,0,0,.1);
```
Comment (rignatus) [inline assets/css/events.css]: You could maybe tweak a little with the shadows as the events' details description part is the same color as the background and I think it becomes exhausting for the eye to follow.

**Review by rignatus — CHANGES_REQUESTED** [review]
Overall a very good improvement. Congrats!  I have added a few comments that should be addressed. I would also suggest adding some more comments as some sections could benefit from them.

File: app.go (hunk 1 above)
Comment (alexdmtr) [reply app.go]: Agreed

File: app.go (hunk 2 above)
Comment (alexdmtr) [reply app.go]: No, I disagree, it makes it less clear what's going on.
What if http.ListenAndServe returns nil (i.e. no-error)?

File: config/config.go (hunk 4 above)
Comment (alexdmtr) [reply config/config.go]: Agreed

File: services/eventService/eventService.go (hunk 8 above)
Comment (alexdmtr) [reply services/eventService/eventService.go:29]: - GetParticipation doesn't include interest
- GetInterestCount is not correct, we are returning a string not a line.
Opinion:
- GetInterestLine should stay

File: services/eventService/eventService.go (hunk 9 above)
Comment (alexdmtr) [reply services/eventService/eventService.go:110]: Agree

File: assets/css/events.css (hunk 3 above)
Comment (alexdmtr) [reply assets/css/events.css]: None.

File: app.go (hunk 2 above)
Comment (rignatus) [reply app.go]: If it returns nil, we just stop the app. 

First of all, we have to understand how `http.ListenAndServe` works and what is the "purpose" of it. When it is called, it blocks and waits forever for clients (so, in our case, it is the "main" thread). If `http.ListenAndServe` returns an `error` or `nil` it means that it has been stopped. From here we have 2 possibilities:
1. If it returned `nil`, we just stop the app (since `log.Fatal()` will not print anything and will do `os.Exit(1)`);
2. If it returned `error`, we log it with `log.Fatal` and exit our app.

In conclusion, if, for whatever the reason, `http.ListenAndServe` stopped, we want to exit as this was our "main" thread of our app.

**Review by rignatus — CHANGES_REQUESTED** [review]
Good improvements on the way we are rendering the templates. 

Other than that, please address all the comments (from a previous/current review) before updating the PQ.

File: config/config.go [hunk 15]
```diff
@@ -11,3 +13,24 @@ func init() {
...
+			missing = append(missing, envvar)
+		}
+	}
+
+	if len(missing) == 0 {
+		return true
+	}
+
+	for _, envvar := range missing {
+		log.Printf("Environment variable %s is not assigned\n", envvar)
```
Comment (rignatus) [inline config/config.go]: Nit: change 
`log.Printf("Environment variable %s is not assigned\n", envvar)` 
to 
`log.Printf("Environment variable %s is not assigned.\n", envvar)`.

File: config/config.go [hunk 16]
```diff
@@ -11,3 +13,24 @@ func init() {
 		log.Fatal("Error loading .env file", err)
 	}
 }
+
+// CheckHaveenv checks if all the env variables given as args exist.
+func CheckHaveenv(vars ...string) bool {
```
Comment (rignatus) [inline config/config.go]: Please fix all references to `os.Getenv` (`databaseService`).

File: services/eventService/eventService.go (hunk 10 above)
Comment (alexdmtr) [reply services/eventService/eventService.go]: Agree, implemented

File: services/eventService/eventService.go (hunk 11 above)
Comment (alexdmtr) [reply services/eventService/eventService.go]: Agree, we should show all environment variables we're lacking.

Fixed! :+1:

File: templates/shared/event-list.tmpl (hunk 12 above)
Comment (alexdmtr) [reply templates/shared/event-list.tmpl]: :+1:

File: config/config.go (hunk 15 above)
Comment (alexdmtr) [reply config/config.go]: Why?

File: config/config.go (hunk 16 above)
Comment (alexdmtr) [reply config/config.go]: Will do :+1:

File: app.go (hunk 2 above)
Comment (alexdmtr) [reply app.go]: I disagree, log.Fatal(http.ListenAndServe) is unclear, since it's unclear that http.ListenAndServe blocks. 
Let's keep it simple.
What if we ever wanted to have something happen after server closes?

File: handlers/events.go (hunk 7 above)
Comment (alexdmtr) [reply handlers/events.go]: :+1:  implemented

File: services/eventService/eventService.go (hunk 9 above)
Comment (alexdmtr) [reply services/eventService/eventService.go:110]: Implemented :+1:

**Review by rignatus — APPROVED** [review]
